[#44841] adding file storage via api endpoint

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/44841
- added API spec for update and creation
- added storages create endpoint
- added properties to storage representer
- added endpoint for creation of oauth credentials
- added endpoint for fetching oauth credentials
- embedded oauth credentials resource in storages
- added representer for oauth client credentials and oauth applications
- added storage request specs

Kharonus committed:
[#44841] added api spec

- https://community.openproject.org/work_packages/44841
- added spec for update and creation

Kharonus committed:
[#44841] adding storages via api v3

- added storages create endpoint
- added properties to representer

Kharonus committed:
[#44841] added api spec for oauth credentials

- added endpoint for creation of oauth credentials
- added endpoint for fetching oauth credentials
- embedded oauth credentials resource in storages

Kharonus committed:
[#44841] rework of api endpoints

- added representer for oauth client credentials and oauth applications

Kharonus committed:
[#44841] added storage request specs for new endpoint

- fixed naming mismatch in API spec
- fixed broken feature spec setup
- fixed rubocop linting
- storage default type is now set by controller

ulferts commented on docs/api/apiv3/paths/oauth_client_credentials.yml:
```diff
@@ -0,0 +1,53 @@
+      required: true
+      schema:
+        type: integer
+      example: 1337
```
This parameter's description is not (no longer?) representing the implementation. Same is in turn true for the endpoint's description. The id is the Oauth client's id. Which is a good thing since otherwise storages would reserve the end point which's usage should be more general. 

I believe at least in the end, the path should be placed under its own tag.

Kharonus replied:
true in all points, fixed.

ulferts commented on docs/api/apiv3/components/schemas/oauth_client_credentials_read_model.yml:
```diff
@@ -0,0 +1,28 @@
+  id: 1337
+  _type: OauthClientCredentials
+  clientId: O5h6WObhMg1Z8IcLHRE3_LMh4jJYmmca2V6OTFSv8DA
+  confidential: true
```
There is no `_self` link here but there should be.

Kharonus replied:
added self link

ulferts commented on docs/api/apiv3/components/schemas/oauth_client_credentials_read_model.yml:
```diff
@@ -0,0 +1,28 @@
+    description: OAuth 2 client id
+  confidential:
+    type: boolean
+    description: true, if OAuth 2 credentials are confidential, false, if no secret is stored
```
My understanding was that the credentials are always confidential. If that holds true, the descriptions should rather be something like:

"true, if a client secret is stored, false otherwise. The secret value is not retrievable."

Kharonus replied:
At least not from an OAuth perspective. It might be a completely valid OAuth 2 connection, with only a client id transported. The client id usually is not considered confidential. The purpose here is merely to verify a specific scope, a client accesses a server.

Kharonus replied:
TBD: should we provide the confidential flag on creation of  `oauth_client_credentials`?

Kharonus replied:
FYI: moved to a later point in time, until we actually need to create non-confidential client credentials

ulferts commented on docs/api/apiv3/components/schemas/oauth_application_read_model.yml:
```diff
@@ -0,0 +1,53 @@
+  id: 1337
+  _type: OauthClientCredentials
+  clientId: O5h6WObhMg1Z8IcLHRE3_LMh4jJYmmca2V6OTFSv8DA
+  confidential: true
```
The example does not match the definition above.

Kharonus replied:
fixed

ulferts commented on docs/api/apiv3/components/schemas/oauth_client_credentials_read_model.yml:
```diff
@@ -0,0 +1,28 @@
+# Schema: OAuthClientCredentialsReadModel
+---
+type: object
```
What about timestamps? Only if this resource is immutable would I not have them included. And even then, I would at least include a `createdAt`

Kharonus replied:
added them. It is indeed immutable in the current design. So we could argue, if we leave `updatedAt` out.

ulferts commented on docs/api/apiv3/components/schemas/storage_read_model.yml:
```diff
@@ -86,6 +96,28 @@ properties:
+
+              - User has role `admin`
+
+              **Resource**: OAuthClientCredentials
```
For somebody not knowing that there is a two way connection, it might help to have a bit more text. Or, add some more general description to the File Link tag which is currently looking like this:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/617519/205942078-715d0a9e-c62b-4eba-a66d-f4eb5ca6cb27.png">

Kharonus replied:
We definitely need a detailed description on the file links tag. :/ Wieland and me just didn't take the time until now.

ulferts commented on docs/api/apiv3/paths/storages.yml:
```diff
@@ -0,0 +1,25 @@
+      content:
+        application/hal+json:
+          schema:
+            $ref: '../components/schemas/oauth_client_credentials_write_model.yml'
```
Without having looked into the implementation yet, I'd say this is the wrong model being returned. Because a `Storage` type should be returned which has the credentials embedded.

Kharonus replied:
fixed

ulferts commented on docs/api/apiv3/components/schemas/storage_read_model.yml:
```diff
@@ -25,6 +25,16 @@ properties:
+      - oauthClientCredentials
+    properties:
+      oauthApplication:
+        $ref: './oauth_client_credentials_read_model.yml'
```
This reference is wrong. It should rather be './oauth_application_read_model.yml`

Kharonus replied:
fixed

ulferts commented on docs/api/apiv3/openapi-spec.yml:
```diff
@@ -774,8 +780,6 @@ components:
     UserPreferencesModel:
       "$ref": "./components/schemas/user_preferences_model.yml"
-    User_PreferenceModel:
-      "$ref": "./components/schemas/user_preference_model.yml"
```
👍 for cleaning those up

Kharonus replied:
I try to reduce everytime I touch the spec the amount of warnings of the validation script by at least 5 ^^

ulferts commented on lib/api/v3/oauth/oauth_client_credentials_api.rb:
```diff
@@ -0,0 +1,44 @@
+  class OAuthClientCredentialsAPI < ::API::OpenProjectAPI
+    resources :oauth_client_credentials do
+      route_param :oauth_client_credentials_id, type: Integer, desc: 'OAuth Client Credentials id' do
+        after_validation do
```
The specification says that admin privileges are. Since this is not nested, currently no permission check is applied.

Kharonus replied:
How do we usually do this?

Kharonus replied:
done

ulferts commented on docs/api/apiv3/paths/oauth_client_credentials.yml:
```diff
@@ -0,0 +1,53 @@
+      description: |-
+        Returned if the storage does not exist or the client does not have sufficient permissions to see it.
+
+        **Required permission:** view file links
```
Since the resource is not nested, this should be altered so that 'Storage' is no longer contained in here. The permission requirement can be  removed.

Kharonus replied:
yes, removed

ulferts commented on docs/api/apiv3/paths/storage.yml:
```diff
@@ -22,7 +22,64 @@ get:
+
+        **Required permission:** view file links
+        
+put:
```
Is this actually a `put` or a `patch`?

Kharonus replied:
it is a `patch`, indeed

ulferts commented on docs/api/apiv3/paths/storages.yml:
```diff
@@ -0,0 +1,25 @@
+    - File links
+  description: |-
+    Creates a storage resource. When creating a storage, a confidential oauth provider is created. The oauth client id
+    and secret are returned in the response.
```
Again, since there are both, the oauth provider as well as the client, the description would benefit from being very specific which's client id and secret are returned.

It might also help, to reference the `POST api/v3/storages/:id/oauth_client_credentials` endpoints in the description or link to some general setup explanation, since a client will have to POST to both endpoints before the setup is complete. 

Just an idea: it might help to have some sort of check (attribute or endpoint) to check if the setup is complete.

Kharonus replied:
Hmm, I hoped, the description is clear. It is meant, that a provider application is created. there are no client credentials inserted yet. This is an implementation detail, hence it should be described comprehensible in this description.

How could a check look like? Should it make a real check, if a connection between NC and OP can be established? Or just try to verify, if all data is available?

Kharonus replied:
amended the description to be more complete and precise

ulferts commented on docs/api/apiv3/components/schemas/storage_read_model.yml:
```diff
@@ -109,3 +141,8 @@ example:
+      href: /api/v3/oauth_credentials/42
+      title: It's no moon (Nextcloud)
+    oauthClientCredentials:
+      href: /api/v3/oauth_credentials/42
```
The hrefs are wrong: oauth_applications and oauth_client_credentials ideally with different ids

Kharonus replied:
fixed

ulferts commented on docs/api/apiv3/paths/storage.yml:
```diff
@@ -22,7 +22,64 @@ get:
+      description: |-
+        Returned if the client does not have sufficient permissions.
+
+        **Required permission:** admin
```
The permission stated to be required for the `404` is incorrect. In fact, it can be deleted completely and the description be shortened to just say: 

'Returned if the storage does not exist.'

Kharonus replied:
I disagree, if a storage exists, but the user does not have the permission to view file links, he should see a `404` when trying to patch or delete this storage.

ulferts reviewed (changes requested):
@Kharonus this is just the first batch of items I noticed. They may be unrefined and wrong as I am missing to complete picture.

ulferts commented on modules/storages/lib/api/v3/oauth_client/oauth_client_credentials_api.rb:
```diff
@@ -0,0 +1,42 @@
+      post &::API::V3::Utilities::Endpoints::Create
+              .new(model: ::OAuthClient,
+                   parse_representer: ::API::V3::OAuth::OAuthClientCredentialsRepresenter,
+                   render_representer: ::API::V3::OAuth::OAuthClientCredentialsRepresenter,
```
There is a parameter for this mapping between internal representation and API name called `api_name`, which is e.g. [used in the membership api](https://github.com/opf/openproject/blob/dev/lib/api/v3/memberships/memberships_api.rb#L42-L48).

Kharonus replied:
maybe I'm using it wrong? Does not work for me :/

Kharonus replied:
we would have to change directory structure to make that work, not needed right now.

ulferts commented on lib/api/v3/oauth/oauth_applications_api.rb:
```diff
@@ -0,0 +1,44 @@
+#++
+
+module API::V3::OAuth
+  class OAuthApplicationsAPI < ::API::OpenProjectAPI
```
The resource is not part of the API documentation.

Kharonus replied:
ouf, good catch!

ulferts commented on lib/api/v3/oauth/oauth_applications_api.rb:
```diff
@@ -0,0 +1,44 @@
+    resources :oauth_applications do
+      route_param :oauth_application_id, type: Integer, desc: 'OAuth application id' do
+        after_validation do
+          @application = ::Doorkeeper::Application.find(params[:oauth_application_id])
```
There is no authorization check - assuming this should be admin only.

Kharonus replied:
same question as above

ulferts commented on lib/api/v3/oauth/oauth_applications_representer.rb:
```diff
@@ -0,0 +1,80 @@
+
+    property :uid, as: :clientId
+
+    property :confidential
```
The concept `confidential` used here has a completely different meaning compared to how it is used in the `OauthClientCredentialsRepresenter` which will ideally be tackled in the docs.

Kharonus replied:
isn't it the exact same meaning? I'm confused. Confidential means, the oauth connection is based not only on a client id, but on a secret, too.

ulferts commented on lib/api/v3/oauth/oauth_applications_representer.rb:
```diff
@@ -0,0 +1,80 @@
+        href: api_v3_paths.storage(represented.integration.id),
+        title: represented.integration.name
+      }
+    end
```
This link surprised me. The oauth application itself is not limited in any way to only be used for the storage. Using a scope to actually do this would be a good approach but is probably too big to tackle right now. But currently, a client can just as well authenticate based on one oauth application as it can based on the other even if only one of those is configured to be "the one linked to a storage". This gets me thinking. There really is no technical link between a storage and an oauth application, is there? 

The only connection I could find after a quick search is that "the one linked to a storage" will be destroyed upon destruction of the storage. 

Even given my currently perceived technical limitations, having this link might be beneficial in the scope of storage so that a client knows which application is somewhat linked to it and this might be a pattern that is recurring. Currently however, that is only a small advantage in my opinion. 

So I would currently tend to rather not have this link. Clients interested in the connection between Oauth Application and Storage will most likely be coming from the storage endpoint anyway.

Kharonus replied:
this link is rendered based on the condition the oauth app has a relation to a storage (or any other integration, which might be... uuurr... another storage :D only storages by now). But the data is based in the oauth application resource, not the storage resource, why I put it here.

Kharonus replied:
fixed

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -94,6 +126,28 @@ def initialize(model, current_user:, embed_links: nil)
+                        link: ->(*) {
+                          return { href: nil } if represented.oauth_client.blank?
+
+                          { href: "/api/v3/oauth_client_credentials/#{represented.oauth_client.id}" }
```
Please use the path helpers here. Otherwise, the paths will be incorrect if the application is running under a nested path e.g.: `https://something.org/openproject`

ulferts replied:
Why is overwriting the default link implementation necessary? Because of this, the link is always displayed even if the current_user is not an admin.

Kharonus replied:
overwrite was necessary:
- on oauth_apps: because we need this service result of the app and not the oauth app model fetched via active relation
- on oauth_client_credentials: because we have no title in this link

reason for apps might be only true for the getter, not the link. Can check that once I solved the other issues.

added apiv3 path helper for the href

Kharonus replied:
reworked representer code here

ulferts commented on modules/storages/lib/api/v3/storages/storages_api.rb:
```diff
@@ -39,6 +39,9 @@ class API::V3::Storages::StoragesAPI < ::API::OpenProjectAPI
   # The `:resources` keyword defines the API namespace -> /api/v3/storages/...
   resources :storages do
+    post &::API::V3::Storages::CreateEndpoint.new(model: ::Storages::Storage).mount
+    # post &::API::V3::Utilities::Endpoints::Create.new(model: ::Storages::Storage).mount
```
Please remove the commented out line of code

Kharonus replied:
removed

ulferts commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -125,6 +129,10 @@ class Engine < ::Rails::Engine
     end
 
+    add_api_path :storage_oauth_client_credentials do |storage_id|
+      "#{root}/storages/#{storage_id}/oauth_client_credentials"
```
It should be possible to also let the paths base on one another since they end up in the same module anyway. So this could be rewritten to:

```
"#{storage(storage_id)}/oauth_client_credentials"
```

same is of course true for the already existing path definitions.

Kharonus replied:
refactored

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -58,16 +80,26 @@ def initialize(model, current_user:, embed_links: nil)
+                            break if fragment['href'].blank?
+
+                            represented.host = fragment['href'].gsub(/\/+$/, '')
+                          }
```
The property is named `origin` here and `host` in the database. This currently leads to errors in the api referencing  the `host` property:

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/617519/206154454-2eafbdfa-d405-4fa3-be8d-01bc002dc2ed.png">

Any chance that we can simply rename `host` to `origin` on the database level?

Kharonus replied:
What is the preferred term? Maybe we should rename `origin` to `host` on API level?

Kharonus replied:
property name mismatch not yet fixed, should be done in a separate PR, IMHO

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -58,16 +80,26 @@ def initialize(model, current_user:, embed_links: nil)
+                            represented.provider_type = case fragment['href']
+                                                        when URN_STORAGE_TYPE_NEXTCLOUD
+                                                          Storages::Storage::PROVIDER_TYPE_NEXTCLOUD
+                                                        end
```
This approach works for now. If ever a form resource for storages is implemented it will fail because invalid values, those not being of a compatible type, will be swallowed silently.

Adding a 

```
else
  fragment['href']
end
```

would work around this and should cause the same error message if necessary.

ulferts replied:
Error messages, if they occur, reference the `providerType` property which does not exist on the API level:

<img width="896" alt="image" src="https://user-images.githubusercontent.com/617519/206157210-337dc9b6-7320-45c7-a131-1c296f49c3e2.png">

The simplest solution would be to rename the property on the database level.

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -58,16 +80,26 @@ def initialize(model, current_user:, embed_links: nil)
-    end
+    link_without_resource :type,
+                          getter: ->(*) {
+                            { href: URN_STORAGE_TYPE_NEXTCLOUD, title: 'Nextcloud' }
```
Every storage will now be mapped to NEXTCLOUD. To avoid surprises in the future, either add a check or map.

Kharonus replied:
we have a couple of those fallbacks in our base. You think it is needed to remove all those hardcoded fallbacks?

Kharonus replied:
fixed

ulferts commented on docs/api/apiv3/components/schemas/storage_write_model.yml:
```diff
@@ -0,0 +1,37 @@
+    origin:
+      href: https://nextcloud.deathstar.rocks
+    type:
+      href: urn:openproject-org:api:v3:storages:nextcloud
```
The implementation uses `urn:openproject-org:api:v3:storages:Nextcloud` ("Nextcloud" with the first letter in uppercase).

ulferts replied:
Alhtough URNs according to my quick research are case sensitive, we could also think about also accepting them case insensitive.

Kharonus replied:
is there any harm in having the check for those URNs case sensitive? TBH, I like it :D

ulferts commented on modules/storages/lib/api/v3/oauth_client/oauth_client_credentials_api.rb:
```diff
@@ -0,0 +1,42 @@
+                   params_modifier: ->(params) do
+                     params.merge(integration: @storage)
+                   end)
+              .mount
```
Sending a request to overwrite the client credentials for a storage, which according to the specs should overwrite the already existing credentials, fails:

<img width="875" alt="image" src="https://user-images.githubusercontent.com/617519/206178121-c6cda6ac-a9f0-4b9a-a038-b310ff05352e.png">

Kharonus replied:
TODO: add request spec

Kharonus replied:
fixed

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -94,6 +126,28 @@ def initialize(model, current_user:, embed_links: nil)
+                        },
+                        link: ->(*) {
+                          {
+                            href: "/api/v3/oauth_applications/#{@oauth_application.id}",
```
Please use the path helpers here. Otherwise, the paths will be incorrect if the application is running under a nested path e.g.: `https://something.org/openproject`

ulferts replied:
Because the default link implementation is overwritten, the link will be displayed even if the current user is not an admin. Is that a conscious decision?

Kharonus replied:
no, this is not intentional, need to fix that

Kharonus replied:
fixed

ulferts commented on modules/storages/spec/lib/api/v3/storages/storages_representer_rendering_spec.rb:
```diff
@@ -68,6 +69,21 @@
+        let(:link) { 'oauthClientCredentials' }
+        let(:href) { nil }
+      end
+    end
```
Some of the paths, leading to the link not being rendered are not covered. e.g having no client credentials defined. 

The effects of this change on `_embedded` are not speced.

Kharonus replied:
added

ulferts commented on modules/storages/spec/requests/api/v3/storages/storages_spec.rb:
```diff
@@ -343,4 +446,61 @@ def query(_)
     end
   end
+
+  describe 'POST /api/v3/storages/:storage_id/oauth_client_credentials' do
```
The GET endpoint is missing request specs. 

If this file gets unwieldy, it could be split up into single files as is e.g. done [for notifications](https://github.com/opf/openproject/tree/dev/spec/requests/api/v3/notifications)

Kharonus replied:
there is no GET on that route

ulferts commented on modules/storages/lib/api/v3/oauth_client/oauth_client_credentials_api.rb:
```diff
@@ -0,0 +1,42 @@
+
+module API::V3::OAuthClient
+  class OAuthClientCredentialsAPI < ::API::OpenProjectAPI
+    resources :oauth_client_credentials do
```
This endpoint lacks request specs.

Kharonus replied:
those are existent

ulferts commented on lib/api/v3/oauth/oauth_client_credentials_api.rb:
```diff
@@ -0,0 +1,44 @@
+          @oauth_client = ::OAuthClient.find(params[:oauth_client_credentials_id])
+        end
+
+        get &::API::V3::Utilities::Endpoints::Show
```
This endpoint lacks request specs.

Kharonus replied:
added

ulferts reviewed (changes requested):
@Kharonus last batch of review feedback. There is a lot in here so I sometimes felt a bit overwhelmed which might have affected the review quality. Maybe we should go over the topics of the review together.

Kharonus committed:
[#44841] added requests specs

- amended API specifications
- delete old oauth client credentials when reposting

Kharonus committed:
[#44841] fixed merge conflict problems

Kharonus committed:
[#44841] fixed representer maps

ulferts commented on docs/api/apiv3/components/schemas/storage_write_model.yml:
```diff
@@ -0,0 +1,37 @@
+properties:
+  name:
+    type: string
+    description: Storage name, if not provided, falls back to a default.
```
The current default only works for one storage. After that, the `name` property de facto becomes required. But when attempting to create a storage without names the second time, a "Name has already been taken." error is displayed. But this error is only displayed if a valid `type` is provided as that determines the default.

Proposal: Append a counter to the default value.

Kharonus replied:
added proposal. please check implementation, if this fits the requirements. it won't work if someone deletes `My Nextcloud 2`, when already any `My Nextcloud 3` exists and then tries to create a storage without name.

not sure, if those edge cases should be supported.

Kharonus replied:
solved with id insertions

ulferts commented on app/services/oauth_clients/create_service.rb:
```diff
@@ -35,5 +35,12 @@
+    def before_perform(params, _service_result)
+      OAuthClient.where(integration: params[:integration]).delete_all
+
+      super
```
Please extend the spec for this service to reflect this deviation from the default behaviour.

Kharonus replied:
amended spec

ulferts commented on docs/api/apiv3/tags/oauth.yml:
```diff
@@ -0,0 +1,4 @@
+---
+description: |-
+  TBD
+name: OAuth 2
```
The file is not referenced in the `tags` section of `openapi-spec.yml`

Kharonus replied:
uf, eagle eye ;)

added

ulferts commented on modules/storages/app/services/storages/storages/create_service.rb:
```diff
@@ -45,6 +45,7 @@ def after_perform(service_call)
         persist_service_result = ::Storages::OAuthApplications::CreateService.new(storage:, user:).call
+        storage.oauth_application = persist_service_result.result if persist_service_result.success?
         service_call.add_dependent!(persist_service_result)
       end
```
A spec for this will reduce the risk of regressions.

Kharonus replied:
amended spec to check for `plaintext_secret`

ulferts commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -35,17 +35,46 @@ module API::V3::Storages
         ::OAuthClients::ConnectionManager.new(user: current_user, oauth_client: model.oauth_client)
 
-      super
+      super(model, current_user:, embed_links:)
```
This change is unnecessary since calling `super` (without args) will just pass on all the args of the method `super` is called in.

Kharonus replied:
reverted

ulferts commented on spec/requests/oauth_clients/oauth_client_credentials_spec.rb:
```diff
@@ -0,0 +1,54 @@
+describe 'API v3 oauth applications resource', content_type: :json do
+  include API::V3::Utilities::PathHelper
+
+  let(:current_user) { create(:admin) }
+  let(:oauth_client_credentials) { create(:oauth_client) }
+
+  before do
+    login_as current_user
```
Not mandatory, but 

```
  let(:current_user) { create(:admin) }

  before do
    login_as current_user
  end
```

can be rewritten to 

```
  current_user { create(:admin) }
```

Kharonus replied:
rewritten

ulferts reviewed (changes requested):
Only small changes needed to get this mergeable, @Kharonus. Thanks for working through my remarks so far.

Admittedly, I only glanced over the specs.

Kharonus committed:
[#44841] some more fixes and regression tests

Kharonus committed:
Merge branch 'dev' into implementation/44841-adding-file-storage-via-api-endpoint

ulferts reviewed (approved):
Well done 👍
